sstables/storage.cc

```diff
@@ -366,13 +366,14 @@ future<> filesystem_storage::create_links_common(const sstable& sst, sstring dst
     // TemporaryTOC is always first, TOC is always last
     auto dst = sstable::filename(dst_dir, sst._schema->ks_name(), sst._schema->cf_name(), sst._version, generation, sst._format, component_type::TemporaryTOC);
     co_await sst.sstable_write_io_check(idempotent_link_file, sst.filename(component_type::TOC), std::move(dst));
-    co_await sst.sstable_write_io_check(sync_directory, dst_dir);
+    auto dir_f = co_await open_checked_directory(sst._write_error_handler, dst_dir);
+    co_await dir_f.flush();
     co_await parallel_for_each(comps, [this, &sst, &dst_dir, generation] (auto p) {
         auto src = sstable::filename(_dir.native(), sst._schema->ks_name(), sst._schema->cf_name(), sst._version, sst._generation, sst._format, p.second);
         auto dst = sstable::filename(dst_dir, sst._schema->ks_name(), sst._schema->cf_name(), sst._version, generation, sst._format, p.second);
         return sst.sstable_write_io_check(idempotent_link_file, std::move(src), std::move(dst));
     });
-    co_await sst.sstable_write_io_check(sync_directory, dst_dir);
+    co_await dir_f.flush();
```

Comment:
what about the `sstable_write_io_check`?
can't we use the dir_wrapper here?

Reply:
> what about the `sstable_write_io_check`?

`dir_f` is created with `open_checked_directory()`, which incorporates `sst._write_error_handler` into the file object via `make_checked_file()`.

This was inspired by commit 8f417b802193 ("sstable: coroutinize seal_sstable", 2022-02-14).

> can't we use the dir_wrapper here?

In my opinion, we can't. (In the commit message, I wrote, *filesystem_storage::create_links_common() runs on directories that generally differ from "_dir", thus, we can't replace its sync_directory() calls with _dir.sync().*)

`create_links_common()` takes `dst_dir` as a parameter. There are four direct call sites:

1. `create_links_common()` overload (3 params)
2. `create_links()`
3. `snapshot()`
4. `move()`. 

All four direct call sites pass directory names that are (a function of) their own input parameters:

- `move()` is called from `change_state()`; the latter passes a pathname that is a child or a sibling of `_dir`.
- `snapshot()` passes in an absolute directory that's unrelated to `_dir`, or a directory that's a (possibly indirect) child of `_dir`.
- AFAICT, `create_links()` is only used from the test suite (`test/boost/sstable_test.cc`, `test/boost/sstable_datafile_test.cc`, `test/lib/sstable_utils.hh`); with directories that are not `_dir` (temp dirs etc).
- the 3-param overload of `create_links_common()` is called by `snapshot()`.

@avikivity [wrote](https://github.com/scylladb/scylladb/pull/19624#discussion_r1672953466), _let's identify the simplest case where we can eliminate the open_; I felt `create_links_common()` was out of scope.

Reply:
okay, let's leave it aside for now.
We may be smarter in the future...

Maybe if the caller would pass a refrrence to dir_wrapper objects for the source and destination directories, we could use "stationary" dir wrappers for permanent locations like the tae base directory or its staging sub directory, and for snapshot, instantiate a dir_wrapper that would live for the duration of snapshot taking (and would be reused to link all sstables into it).

I think there should be another use case for intra-node tablet migration when we move sstables between shards instead of streaming them, but maybe it doesn't use create_links?
Cc @raphaelsc

Reply:
I am not following the discussion, but yes, intra-node migration does rely on hard linking.

Reply:
> okay, let's leave it aside for now.
> We may be smarter in the future...

Thanks, I'm resolving this thread then; it's hard to contain them all.